Distiller-server/src/main/java/com/mapr/distiller/server/metricactions/SystemCpuMetricAction.java

```diff
@@ -0,0 +1,162 @@
+import com.mapr.distiller.server.queues.RecordQueueManager;
+import com.mapr.distiller.server.recordtypes.Record;
+import com.mapr.distiller.server.recordtypes.SystemCpuRecord;
+import com.mapr.distiller.server.utils.Constants;
+import com.mapr.distiller.server.utils.MetricConfig;
+
+public class SystemCpuMetricAction extends MetricAction implements
+		MetricsSelectable {
+	Object object = new Object();
+
+	private RecordQueue inputQueue;
+	private RecordQueue outputQueue;
+	private String recordType;
+
+	private SystemCpuRecordProcessor recordProcessor;
+
+	private String aggregationType;
+	private Map<String, String> aggregationMap;
+
+	private boolean shouldPersist;
+
+	public SystemCpuMetricAction(String id, String recordType,
+			String aggregationType, Map<String, String> aggregationMap,
+			RecordQueue inputQueue, RecordQueue outputQueue) {
+		super(id);
+		this.recordType = recordType;
+		this.aggregationType = aggregationType;
+		this.aggregationMap = aggregationMap;
+
+		this.inputQueue = inputQueue;
+		this.outputQueue = outputQueue;
+
+		setGathericMetric(true);
+	}
+
+	// We also need one more argument which has a RecordQueue map to their names
+	public static SystemCpuMetricAction getInstance(MetricConfig metricConfig,
+			RecordQueueManager queueManager) {
+		return new SystemCpuMetricAction(metricConfig.getId(),
+				metricConfig.getRecordType(),
+				metricConfig.getAggregationType(),
+				metricConfig.getAggregationMap(),
+				queueManager.getQueue(metricConfig.getInputQueue()),
+				queueManager.getQueue(metricConfig.getOutputQueue()));
+	}
+
+	@Override
+	public void run() {
+		System.out.println("****" + this.recordType + "****");
+		while (!Thread.interrupted()) {
+			if (isGathericMetric()) {
+				selectSequentialRecords();
+				/*try {
+					Thread.sleep(1000);
+				} catch (InterruptedException e) {
+					System.out.println("Thread got interrupted - Going down");
+					break;
+				}*/
+			}
+
+			else {
+				try {
+					synchronized (object) {
+						object.wait();
+					}
+				} catch (InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+			}
+		}
+
+		System.out.println("Exiting Metric action due to kill action");
+	}
+
+	@Override
+	public void suspend() throws InterruptedException {
+		System.out.println("Stopping metric with id = " + id);
+		if (isGathericMetric) {
+			isGathericMetric = false;
+		}
+
+		else {
+			System.out.println("Already suspended metric " + id);
+		}
+	}
+
+	@Override
+	public void resume() {
+		synchronized (object) {
+			if (!isGathericMetric) {
+				System.out.println("Resuming metric with id = " + id);
+				isGathericMetric = true;
+				object.notifyAll();
+			}
+
+			else {
+				System.out.println("Already running Metric " + id);
+			}
+		}
+	}
+
+	@Override
+	public void kill() {
+		System.out.println("Kill metric with id = " + id);
+		isGathericMetric = false;
+	}
+
+	@Override
+	public void selectSequentialRecords() {
```

Comment:
Hi Venkat,

What if we put the thresholdable/moving averageable/etc. methods direclty
in Record?  We could have it default to producing a null record (for moving
average) or false (for thresholdable), and then each Record extending class
would override.  That would the coordination framework has the methods
there no matter what and if a user tries to apply a type of processing
against a record that doesn't support it then they wouldn't get anything
out of it.

On Wed, Aug 5, 2015 at 3:51 PM, Venkata krishnan Sowrirajan <
notifications@github.com> wrote:

> In
> Distiller-server/src/main/java/com/mapr/distiller/server/metricactions/SystemCpuMetricAction.java
> <https://github.com/vsowrirajan/Distiller/pull/12#discussion_r36364270>:
>
> > +			}
> > +
> > +			else {
> > +				System.out.println("Already running Metric " + id);
> > +			}
> > +		}
> > +	}
> > +
> > +	@Override
> > +	public void kill() {
> > +		System.out.println("Kill metric with id = " + id);
> > +		isGathericMetric = false;
> > +	}
> > +
> > +	@Override
> > +	public void selectSequentialRecords() {
>
> I was thinking about that, but the thing is MetricSelectable is an
> interface which provides those methods. We are not implementing
> MetricSelectable in the base class MetricAction (MetricAction.java) rather
> we are implementing it in the record type metric action (eg:
> SystemCpuMetricAction) itself. So I think we need that to be on their own
> record type metric actions (eg: SystemCpuMetricAction). Though the
> selectSequentialRecords is going to be logically same for every RecordType.
> Each metric action know which type of record processor (eg:
> SystemCpuRecordProcessor) to use. We are having a RecordProcessor
> (RecordProcessor.java) class right now which is obsolete. Its not how we
> have designed the record processors (eg: SystemCpuRecordProcessor).
>
> Its more of each record type metric action will have its own record type
> record processors (for eg: SystemCpuMetricAction =>
> SystemCpuRecordProcessor). This is how I thought of initially, so I am not
> sure how can we put those "selector" methods in one place and use it with
> the processors.
>
> We will not merge this for some time. I will think about this for a while
> to see if I can come up with a better approach or not. If you have any
> other suggestions, please pass on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vsowrirajan/Distiller/pull/12/files#r36364270>.
>